Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-        }
-    });
 
     resetReadyForMoreSample();
 }
 
+void VideoMediaSampleRenderer::decodedFrameAvailable(RetainPtr<CMSampleBufferRef>&& sample)
+{
+    assertIsCurrent(m_workQueue);
+
+    assignResourceOwner(sample.get());
+
+    if (m_timebase) {
+        purgeDecodedSampleQueue();
+        PAL::CMBufferQueueEnqueue(ensureDecodedSampleQueue(), sample.get());
+    }
+
+    if (m_renderer)
+        [m_renderer enqueueSampleBuffer:sample.get()];
+    else if (m_displayLayer) {
+        callOnMainThread([weakThis = ThreadSafeWeakPtr { *this }, sample = sample] {
+            ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+            if (RefPtr protectedThis = weakThis.get())
+                [protectedThis->m_displayLayer enqueueSampleBuffer:sample.get()];
+            ALLOW_DEPRECATED_DECLARATIONS_END
+        });
+    }
+}
+
+void VideoMediaSampleRenderer::flushCompressedSampleQueue()
+{
+    assertIsCurrent(m_workQueue);
+    if (!m_compressedSampleQueue)
+        return;
+
+    PAL::CMBufferQueueReset(m_compressedSampleQueue.get());
+    PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(m_timebase.get(), m_timerSource.get(), PAL::kCMTimeInvalid, 0);
+}
+
+void VideoMediaSampleRenderer::flushDecodedSampleQueue()
+{
+    assertIsCurrent(m_workQueue);
+    if (!m_decodedSampleQueue)
+        return;
+
+    PAL::CMBufferQueueReset(m_decodedSampleQueue.get());
+    PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(m_timebase.get(), m_timerSource.get(), PAL::kCMTimeInvalid, 0);
+}
+
+void VideoMediaSampleRenderer::purgeDecodedSampleQueue()
+{
+    assertIsCurrent(m_workQueue);
+    if (!m_decodedSampleQueue)
+        return;
+
+    if (!m_timebase)
+        return;
+
+    CMTime currentTime = PAL::CMTimebaseGetTime(m_timebase.get());
+    CMTime nextPurgeTime = PAL::kCMTimeInvalid;
+
+    while (auto nextSample = (CMSampleBufferRef)const_cast<void*>(PAL::CMBufferQueueGetHead(m_decodedSampleQueue.get()))) {
+        CMTime presentationTime = PAL::CMSampleBufferGetOutputPresentationTimeStamp(nextSample);
+        CMTime duration = PAL::CMSampleBufferGetOutputDuration(nextSample);
+        CMTime presentationEndTime = PAL::CMTimeAdd(presentationTime, duration);
+        if (PAL::CMTimeCompare(presentationEndTime, currentTime) >= 0) {
+            nextPurgeTime = presentationEndTime;
+            break;
+        }
+
+        RetainPtr sampleToBePurged = adoptCF(PAL::CMBufferQueueDequeueAndRetain(m_decodedSampleQueue.get()));
+        sampleToBePurged = nil;
+    }
+
+    if (!CMTIME_IS_VALID(nextPurgeTime))
+        return;
+
+    PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(m_timebase.get(), m_timerSource.get(), nextPurgeTime, 0);
+}
+
+CMBufferQueueRef VideoMediaSampleRenderer::ensureCompressedSampleQueue()
+{
+    assertIsCurrent(m_workQueue);
+    if (!m_compressedSampleQueue)
+        m_compressedSampleQueue = WebCoreDecompressionSession::createBufferQueue();
+    return m_compressedSampleQueue.get();
```

Comment:
should be returning RetainPtr<CMBufferQueueRef> to align with new coding safety style